resources/lib/router.py

```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+# Crunchyroll
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+from typing import Optional, Dict
+
+PARAMETER_ROUTE_MODE: str = "__parameter"
+
+# A route is formatted with {parameters}, its linked value will be the "mode" parameter.
+# If value is PARAMETER_ROUTE_MODE, the "mode" parameter will be taken from the url.
+plugin_routes: dict = {
+    "/menu/{mode}": PARAMETER_ROUTE_MODE,
+    "/menu/{mode}/offset/{offset}": PARAMETER_ROUTE_MODE,
+    "/menu/{mode}/{genre}": PARAMETER_ROUTE_MODE,
+    "/menu/{mode}/{genre}/offset/{offset}": PARAMETER_ROUTE_MODE,
+    "/menu/{mode}/{genre}/category/{category_filter}": PARAMETER_ROUTE_MODE,
+    "/menu/{mode}/{genre}/category/{category_filter}/offset/{offset}": PARAMETER_ROUTE_MODE,
+    "/menu/{mode}/{genre}/season/{season_filter}": PARAMETER_ROUTE_MODE,
+    "/menu/{mode}/{genre}/season/{season_filter}/offset/{offset}": PARAMETER_ROUTE_MODE,
+    "/series/{series_id}": "series",
+    "/series/{series_id}/{collection_id}": "episodes",
+    "/series/{series_id}/{collection_id}/offset/{offset}": "episodes",
+    "/video/{series_id}/{episode_id}/{stream_id}": "videoplay",
+    "/video/{episode_id}/{stream_id}": "videoplay"
+}
```

Comment:
Move to this format, it will avoid usage of magical mode value (`PARAMETER_ROUTE_MODE`) and provide a way to designate a specific route.

```python
{
  "route_name": {
    "url": "/menu/{mode}"
  },
  "series_view": {
    "url": "/series/{series_id}",
    "mode": "series"
  },
  # ...
}
```